### PR TITLE
feat(provider/lambda) : adding support for aws lambda

### PR DESCRIPTION
**Pull request by anshrma**

### CODE REVIEW

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java [hunk 1]
```diff
@@ -87,4 +87,15 @@
...
+  // AWS Lambda Operations
+
+  public static final String INVOKE_AWSLAMBDA = "invokeAwsLambda";
+  public static final String CREATE_AWSLAMBDA = "createAwsLambda";
+  public static final String UPDATE_AWSLAMBDA_CONFIGURATION = "updateAwsLambdaConfiguration";
+  public static final String UPDATE_AWSLAMBDA_CODE = "updateAwsLambdaCode";
+  public static final String DELETE_AWSLAMBDA = "deleteAwsLambda";
+  public static final String UPSERT_AWSLAMBDA_ALIAS = "upsertAwsLambdaAlias";
+  public static final String UPSERT_AWSLAMBDA_EVENTMAPPING = "upsertAwsLambdaEventMapping";
+  public static final String DELETE_AWSLAMBDA_EVENTMAPPING = "deleteAwsLambdaEventMapping";
```
Comment (robzienert) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: Do we have any appetite for making these cloud agnostic? E.g. `invokeFunction`, `createFunction`, and so-on.

@spinnaker/reviewers

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+public class AwsLambdaBasicController {
+
+  AwsLambdaProvider awsLambdaProvider;
+
+  @Autowired
+  public AwsLambdaBasicController(AwsLambdaProvider awsLambdaProvider) {
+    this.awsLambdaProvider = awsLambdaProvider;
+  }
+
+
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java]: Extra space(s) in this file.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaDeleteEventMappingAtomicOperationConverter.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  @Override
+  public UpsertLambdaEventMappingDescription convertDescription(Map input) {
+    UpsertLambdaEventMappingDescription converted = getObjectMapper().convertValue(input, UpsertLambdaEventMappingDescription.class);
+    converted.setCredentials(getCredentialsObject(input.get("credentials").toString()));
+
+    return converted;
+  }
+
+  }
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaDeleteEventMappingAtomicOperationConverter.java]: Mind the formatting.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaUpsertEventMappingAtomicOperationConverter.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  @Override
+  public UpsertLambdaEventMappingDescription convertDescription(Map input) {
+    UpsertLambdaEventMappingDescription converted = getObjectMapper().convertValue(input, UpsertLambdaEventMappingDescription.class);
+    converted.setCredentials(getCredentialsObject(input.get("credentials").toString()));
+
+    return converted;
+  }
+
+  }
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaUpsertEventMappingAtomicOperationConverter.java]: Formatting.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/description/UpdateLambdaCodeDescription.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.List;
+import java.util.Map;
+
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class UpdateLambdaCodeDescription extends AbstractAwsLambdaDescription {
+  String s3bucket;
+  String s3key;
+  Boolean publish;
+  }
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/description/UpdateLambdaCodeDescription.java]: Formatting.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaAtomicOperation.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import com.amazonaws.services.lambda.AWSLambda;
+import com.amazonaws.services.lambda.model.DeleteFunctionRequest;
+import com.amazonaws.services.lambda.model.DeleteFunctionResult;
+import com.netflix.spinnaker.clouddriver.lambda.cache.model.AwsLambdaCacheModel;
+import com.netflix.spinnaker.clouddriver.lambda.deploy.description.DeleteLambdaDescription;
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation;
+
+import java.util.List;
+
+;
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaAtomicOperation.java]: ???

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaEventSourceAtomicOperation.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+
+  public DeleteLambdaEventSourceAtomicOperation(UpsertLambdaEventMappingDescription description) {
+    super(description, "DELETE_LAMBDA_FUNCTION_EVENT_MAPPING");
+  }
+
+  @Override
+  public Object operate(List priorOutputs) {
+
+    AwsLambdaCacheModel cache = awsLambdaProvider.getAwsLambdaFunction(description.getProperty("application").toString(),description.getProperty("region").toString(),description.getAccount());
```
Comment (robzienert) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaEventSourceAtomicOperation.java]: Please keep column width at or below 120 characters.

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java (hunk 1 above)
Comment (lwander) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: I was hoping to see a design doc before this PR dropped -- could you write something up @anshrma? https://www.spinnaker.io/community/contributing/submitting/#before-you-begin

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/cache/model/AwsLambdaCacheModel.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.lambda.cache.model;
+import com.amazonaws.services.lambda.model.FunctionConfiguration;
+import com.amazonaws.services.lambda.model.AliasConfiguration;
+import com.amazonaws.services.lambda.model.EventSourceMappingConfiguration;
+
+import java.awt.*;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/cache/model/AwsLambdaCacheModel.java]: Probably don't want this

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  public Collection<AwsLambdaCacheModel> getFunction(@RequestParam(value="functionname",required = false) String functionname,
+                                                     @RequestParam(value="accountname",required = false) String accountname,
+                                                     @RequestParam(value="region",required = false) String region) {
+    if (functionname == null)
+      return awsLambdaProvider.getAllAwsLambdaFunctions();
+    else {
+      ArrayList<AwsLambdaCacheModel> cache = new ArrayList<>();
+      cache.add(awsLambdaProvider.getAwsLambdaFunction(functionname,region,accountname));
+      return cache;
+    }
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java]: Minor style issue, I would rather go with:

    if (functionname == null) {
      return awsLambdaProvider.getAllAwsLambdaFunctions();
    }

    ArrayList<AwsLambdaCacheModel> cache = new ArrayList<>();
    cache.add(awsLambdaProvider.getAwsLambdaFunction(functionname,region,accountname));
    return cache;

Comment (dreynaud) [reply clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java]: So looking up with no `functionname` means returning all functions in all regions and all accounts even if a region and an account were passed?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaCreateAtomicOperationConverter.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  @Override
+  public AtomicOperation convertOperation(Map input) {
+    return new CreateLambdaAtomicOperation(convertDescription(input));
+  }
+
+  @Override
+  public CreateLambdaDescription convertDescription(Map input) {
+    CreateLambdaDescription converted = getObjectMapper().convertValue(input, CreateLambdaDescription.class);
+    converted.setCredentials(getCredentialsObject(input.get("credentials").toString()));
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/converters/AwsLambdaCreateAtomicOperationConverter.java]: this line is repeated 8 times, would it make sense to factor it into a method?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/description/UpsertLambdaAliasDescription.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class UpsertLambdaAliasDescription extends AbstractAwsLambdaDescription {
+  String aliasdescription;
+  String majorfunctionversion;
+  String aliasname;
+  String minorfunctionversion;
+  Double weighttominorfunctionversion;
+
+  }
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/description/UpsertLambdaAliasDescription.java]: I know what @robzienert is going to say here

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/AbstractAwsLambdaAtomicOperation.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    this.description = description;
+    this.basePhase = basePhase;
+
+  }
+
+  private static Task getTask() {
+    return TaskRepository.threadLocalTask.get();
+  }
+
+//  String getCluster(String service, String account) {
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/AbstractAwsLambdaAtomicOperation.java]: leftover from something?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/AbstractAwsLambdaAtomicOperation.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    NetflixAmazonCredentials credentialAccount = description.getCredentials();
+    return amazonClientProvider.getAmazonLambda(credentialAccount,region);
+  }
+
+  protected String getRegion() {
+    return description.getRegion();
+  }
+
+  AmazonCredentials getCredentials() {
+    return (AmazonCredentials) accountCredentialsProvider.getCredentials(description.getCredentialAccount());
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/AbstractAwsLambdaAtomicOperation.java]: any chance that the cast might fail?

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java (hunk 1 above)
Comment (duftler) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: Making the operation names provider-agnostic makes sense to me. Resolved is done in the context of a given provider when executed, so don't see any downside.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaEventSourceAtomicOperation.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    for (Object x : eventmappings){
+      /*
+      anshrma@amazon.com : This is not ideal, but intentional. Ideally, we know that x is not POJO, but AliasConfiguration. However,
+      if we iterate though AliasConfiguration instead, we get an exception java.util.LinkedHashMap cannot be cast to com.amazonaws.services.lambda.model.AliasConfiguration.
+      Hence this workaround
+      */
+
+      HashMap<String,String> eventmap = (HashMap<String, String>) x;
+      if (eventmap.get("eventSourceArn").toLowerCase().equals(description.getProperty("eventsourcearn").toString().toLowerCase())) {
+        flagexists = true;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/DeleteLambdaEventSourceAtomicOperation.java]: do we want to break out of the loop at this point?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/UpsertLambdaAliasAtomicOperation.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+    for (Object x : aliases){
+      /*
+      anshrma@amazon.com : This is not ideal, but intentional. Ideally, we know that x is not POJO, but AliasConfiguration. However,
+      if we iterate though AliasConfiguration instead, we get an exception java.util.LinkedHashMap cannot be cast to com.amazonaws.services.lambda.model.AliasConfiguration.
+      Hence this workaround
+      */
+
+      HashMap<String,String> aliasmap = (HashMap<String, String>) x;
+      if (aliasmap.get("name").toLowerCase().equals(description.getProperty("aliasname").toString().toLowerCase())) {
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/UpsertLambdaAliasAtomicOperation.java]: there is some serious code duplication with that loop, can it be factored out?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/UpsertLambdaEventSourceAtomicOperation.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+public class UpsertLambdaEventSourceAtomicOperation extends AbstractAwsLambdaAtomicOperation<UpsertLambdaEventMappingDescription, Object> implements AtomicOperation<Object> {
+
+
+  public UpsertLambdaEventSourceAtomicOperation(UpsertLambdaEventMappingDescription description) {
+    super(description, "UPSERT_LAMBDA_FUNCTION_EVENT_MAPPING");
+  }
+
+  @Override
+  public Object operate(List priorOutputs) {
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/deploy/ops/UpsertLambdaEventSourceAtomicOperation.java]: I don't think I've seen that pattern of returning `Object` anywhere else in clouddriver, is there any way to be more specific?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+import com.amazonaws.services.lambda.model.*;
+import com.netflix.spinnaker.cats.agent.*;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.cache.DefaultCacheData;
+import com.netflix.spinnaker.cats.provider.ProviderCache;
+import com.netflix.spinnaker.clouddriver.aws.provider.AwsInfrastructureProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.NetflixAmazonCredentials;
+import com.netflix.spinnaker.clouddriver.lambda.cache.Keys;
+import groovy.util.logging.Slf4j;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: I'm surprised it works with the groovy `Slf4j`. Did you mean the lombok `Slf4j`?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+import static com.netflix.spinnaker.clouddriver.lambda.cache.Keys.Namespace.IAM_ROLE;
+
+@Slf4j
+public class AmazonLambdaFunctionCachingAgent implements CachingAgent, AccountAware {
+
+  private final AmazonClientProvider amazonClientProvider;
+  private final NetflixAmazonCredentials account;
+  private final String region;
+  private static final Set<AgentDataType> types = Collections.unmodifiableSet(DefaultGroovyMethods.asType(new ArrayList<AgentDataType>(Arrays.asList(AUTHORITATIVE.forType(LAMBDA_NAME.ns))), Set.class));
+  private final Logger log = LoggerFactory.getLogger(getClass());
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: probably wouldn't need that line with the lombok `@Slf4j`

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+      attributes.put("runtime",x.getRuntime());
+      attributes.put("revisionids",lstFunctionRevisions(x.getFunctionArn()));
+      attributes.put("aliasconfiguration",listAliasConfiguration(x.getFunctionArn()));
+      attributes.put("eventsourcemappings",listEventSourceMappingConfiguration(x.getFunctionArn()));
+
+      DefaultCacheData item = new DefaultCacheData(Keys.getLambdaFunctionKey(account.getName(),region,x.getFunctionName()),attributes,Collections.emptyMap());
+      data.add(item);
+    }
+
+    log.info("Caching " + String.valueOf(data.size()) + " items in " + getAgentType());
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: log.info("Caching {} items in {}", data.size(), getAgentType());

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+
+    } while (nextMarker != null && nextMarker.length() != 0);
+    
+    Collection<CacheData> data = new LinkedList<>();
+    for (FunctionConfiguration x : lstFunction) {
+      //AwsLambdaCacheModel objLambdaCacheModel = new AwsLambdaCacheModel();
+//      objLambdaCacheModel.setFunctionName(x.getFunctionName());
+//      objLambdaCacheModel.setFunctionArn(x.getFunctionArn());
+//      objLambdaCacheModel.setMemorySize(x.getMemorySize());
+//      objLambdaCacheModel.setRevisionids(lstFunctionRevisions(x.getFunctionArn()));
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: please remove if this is not needed anymore

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+import static com.netflix.spinnaker.clouddriver.lambda.cache.Keys.Namespace.LAMBDA_NAME;
+import static com.netflix.spinnaker.clouddriver.lambda.cache.Keys.Namespace.IAM_ROLE;
+
+@Slf4j
+public class AmazonLambdaFunctionCachingAgent implements CachingAgent, AccountAware {
+
+  private final AmazonClientProvider amazonClientProvider;
+  private final NetflixAmazonCredentials account;
+  private final String region;
+  private static final Set<AgentDataType> types = Collections.unmodifiableSet(DefaultGroovyMethods.asType(new ArrayList<AgentDataType>(Arrays.asList(AUTHORITATIVE.forType(LAMBDA_NAME.ns))), Set.class));
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: Would this work instead?

    private static final Set<AgentDataType> types = Collections.singleton(AUTHORITATIVE.forType(LAMBDA_NAME.ns));

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @Override
+  public Collection<AgentDataType> getProvidedDataTypes() {
+    return types;
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    log.info("Describing items in " + getAgentType());
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/AmazonLambdaFunctionCachingAgent.java]: log.info("Describing items in {}", getAgentType());

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/IamPolicyReader.java [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class IamPolicyReader {
+  private static final Logger logger = LoggerFactory.getLogger(IamPolicyReader.class);
+
+  ObjectMapper mapper;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/IamPolicyReader.java]: could probably be `private final`

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/view/AwsLambdaAccountMapper.java [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class AwsLambdaAccountMapper {
+
+  final AccountCredentialsProvider accountCredentialsProvider;
+  final Map<String, NetflixAssumeRoleAwsLambdaCredentials> awslambdaCredentialsMap;
+  final Map<String, NetflixAmazonCredentials> awsCredentialsMap;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/view/AwsLambdaAccountMapper.java]: all fields could be final?

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+import com.netflix.spinnaker.clouddriver.lambda.provider.view.AwsLambdaProvider;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
```
Comment (dreynaud) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/controllers/AwsLambdaBasicController.java]: unused?

**Pull request merged**